crates/walrus-service/src/event/event_processor/processor.rs

```diff
@@ -327,6 +408,70 @@ impl EventProcessor {
         Ok(())
     }
 
+    fn start_tailing_task(
+        &self,
+        cancel_token: CancellationToken,
+        coordination_state: Arc<CatchupCoordinationState>,
+    ) -> JoinHandle<Result<()>> {
+        let processor = self.clone();
+        coordination_state.mark_tailing_started();
+        let coordination_state = coordination_state.clone();
+        tokio::spawn(async move {
+            tracing::info!("Starting tailing task");
+            let result = processor.start_tailing_checkpoints(cancel_token).await;
+            tracing::info!("Tailing task exited");
+            coordination_state.notify_tailing_stopped();
+            result
+        })
+    }
+
+    /// Checkpoint tailing with message-based coordination for catchup
+    async fn start_tailing_checkpoints_with_restart_support(
+        &self,
+        cancel_token: CancellationToken,
+        coordination_state: Arc<CatchupCoordinationState>,
+        mut coordination_rx: mpsc::UnboundedReceiver<CoordinationMessage>,
+    ) -> Result<()> {
+        let mut child_cancel_token = cancel_token.child_token();
+
+        let mut tailing_task: Option<JoinHandle<Result<()>>> =
+            Some(self.start_tailing_task(child_cancel_token.clone(), coordination_state.clone()));
+        loop {
+            tokio::select! {
+                msg = coordination_rx.recv() => {
+                    match msg {
+                        Some(CoordinationMessage::StopCheckpointTailing) => {
+                            child_cancel_token.cancel();
+                            match tailing_task.take().expect("tailing task not running").await {
```

Comment:
Suggestion: instead of panic, can we log an error if tailing_task is None and continue? At least this can be true today since catchup_active is not set.

Reply:
This is handled now.